src/main/java/io/fabric8/jenkins/openshiftsync/BuildSyncRunListener.java

```diff
@@ -292,13 +313,35 @@ private void upsertBuild(Run run, RunExt wfRunExt) {
 				logger.log(Level.FINE, "upsertBuild", t);
 		}
 
+		Map<String,BlueRunResult> blueRunResults = new HashMap<String,BlueRunResult>();
+		if (blueRun != null) {
+	        Iterator<BluePipelineNode> iter = blueRun.getNodes().iterator();
+	        while (iter.hasNext()) {
+	            BluePipelineNode node = iter.next();
+	            if (node != null) {
+	                blueRunResults.put(node.getDisplayName(), node.getResult());
+	            }
+	        }
+		}
 		boolean pendingInput = false;
 		if (!wfRunExt.get_links().self.href.matches("^https?://.*$")) {
 			wfRunExt.get_links().self.setHref(joinPaths(rootUrl, wfRunExt.get_links().self.href));
 		}
 		int newNumStages = wfRunExt.getStages().size();
 		int newNumFlowNodes = 0;
+		Map<String,StageNodeExt> validStages = new HashMap<String,StageNodeExt>();
 		for (StageNodeExt stage : wfRunExt.getStages()) {
+		    // the StatusExt.getStatus() cannot be trusted for declarative pipeline;
+		    // for example, skipped steps/stages will be marked as complete;
+		    // we leverage the blue ocean state machine to determine this
+		    BlueRunResult result = blueRunResults.get(stage.getName());
+		    if (result != null && result == BlueRunResult.NOT_BUILT) {
+		        logger.info("skipping stage " + stage.getName() + " for the status JSON for pipeline run " + run.getDisplayName() + 
+		                " because it was not executed (most likely because of a failure in another stage)");
+		        continue;
+		    }
+		    validStages.put(stage.getName(), stage);
```

Comment:
nit: this block isn't aligned w/ the rest of the code in this for-loop.  needs formatting?